include/irr/asset/filters/CSummedAreaTableImageFilter.h

```diff
@@ -192,40 +178,82 @@ class CSummedAreaTableImageFilter : public CMatchedSizeInOutImageFilterCommon, p
 					const auto blockDims = asset::getBlockDimensions(state->inImage->getCreationParameters().format);
+					static constexpr uint8_t maxPlanes = 4;
+
+					/*
+						Make sure we are able to move as (+ 1) in a certain plane,
+						otherwise memory leaks may occur
+					*/
+
+					bool is2DAndBellow = state->inImage->getCreationParameters().type == IImage::ET_2D;
+					bool is3DAndBellow = state->inImage->getCreationParameters().type == IImage::ET_3D;
+					const core::vectorSIMDu32 limit(1, is2DAndBellow, is3DAndBellow);
+					const core::vectorSIMDu32 movingExclusiveVector = limit, movingOnYZorXZorXYCheckingVector = limit;
 					
 					auto decode = [&](uint32_t readBlockArrayOffset, core::vectorSIMDu32 readBlockPos) -> void
 					{
-						auto newReadBlockPos = (readBlockPos - decltype(readBlockPos)(state->inOffset.x, state->inOffset.y, state->inOffset.z)); // not sure
+						auto newReadBlockPos = (readBlockPos - decltype(readBlockPos)(state->inOffset.x, state->inOffset.y, state->inOffset.z)); 
+						core::vectorSIMDu32 localOutPos = newReadBlockPos * blockDims;
 
-						auto localOutPos = newReadBlockPos * blockDims;
 						auto* inDataAdress = inData + readBlockArrayOffset;
-						const size_t scratchOffset = asset::IImage::SBufferCopy::getLocalByteOffset(core::vector3du32_SIMD(localOutPos.x, localOutPos.y, localOutPos.z), scratchByteStrides);
-						decodeType* outDataAdress = (decodeType*)(reinterpret_cast<uint8_t*>(scratchMemory) + scratchOffset);
-
-						constexpr uint8_t maxPlanes = 4;
 						const void* inSourcePixels[maxPlanes] = { inDataAdress, nullptr, nullptr, nullptr };
 
-						for (auto blockY = 0u; blockY < blockDims.y; blockY++)
-							for (auto blockX = 0u; blockX < blockDims.x; blockX++)
-							{
-								decodeType decodeBuffer[maxChannels] = {};
+						if constexpr (ExclusiveMode)
+						{
+							auto movedLocalOutPos = localOutPos + movingExclusiveVector;
+							const auto isSatMemorySafe = (movedLocalOutPos < core::vectorSIMDu32(state->extent.width, state->extent.height, state->extent.depth, movedLocalOutPos.w + 1)); // force true on .w
+							const auto doesItMoveOnYZorXZorXY = (localOutPos < movingOnYZorXZorXYCheckingVector);
 
-								asset::decodePixelsRuntime(inFormat, inSourcePixels, decodeBuffer, blockX, blockY);
-								memcpy(outDataAdress, decodeBuffer, scratchTexelByteSize);
+							if (isSatMemorySafe.all())
+							{
+								for (auto blockY = 0u; blockY < blockDims.y; blockY++)
+									for (auto blockX = 0u; blockX < blockDims.x; blockX++)
+									{
+										decodeType decodeBuffer[maxChannels] = {};
+
+										if (doesItMoveOnYZorXZorXY.any())
+										{
+											const size_t offset = asset::IImage::SBufferCopy::getLocalByteOffset(core::vector3du32_SIMD(localOutPos.x + blockX, localOutPos.y + blockY, localOutPos.z), scratchByteStrides);
+											memcpy(reinterpret_cast<uint8_t*>(scratchMemory) + offset, decodeBuffer, scratchTexelByteSize);
+										}
+										else
```

Comment:
this should be done outside the decode loop (in a separate loop to clear the XY, XZ, YZ planes), this currently breaks loop unrolling and automatic SIMD and readability

Reply:
corrected in following commit